sys/kern/vm_amap.c

```diff
@@ -153,14 +196,60 @@ void vm_amap_remove_pages(vm_aref_t aref, size_t start, size_t nslots) {
 }
 
 void vm_amap_hold(vm_amap_t *amap) {
-  refcnt_acquire(&amap->ref_cnt);
+  SCOPED_MTX_LOCK(&amap->mtx);
+  amap->ref_cnt++;
 }
 
 void vm_amap_drop(vm_amap_t *amap) {
-  if (refcnt_release(&amap->ref_cnt)) {
-    vm_amap_remove_pages_unlocked(amap, 0, amap->slots);
-    kfree(M_AMAP, amap->pg_list);
-    kfree(M_AMAP, amap->pg_bitmap);
-    pool_free(P_VM_AMAP_STRUCT, amap);
+  WITH_MTX_LOCK (&amap->mtx) {
+    amap->ref_cnt--;
+    if (amap->ref_cnt >= 1)
+      return;
+  }
+  vm_amap_remove_pages_unlocked(amap, 0, amap->slots);
+  kfree(M_AMAP, amap->anon_list);
+  kfree(M_AMAP, amap->anon_bitmap);
+  pool_free(P_VM_AMAP_STRUCT, amap);
+}
+
+static vm_anon_t *alloc_empty_anon(void) {
+  vm_anon_t *anon = pool_alloc(P_VM_ANON_STRUCT, M_WAITOK);
+  anon->ref_cnt = 1;
+  anon->page = NULL;
+  mtx_init(&anon->mtx, MTX_SLEEP);
+  return anon;
+}
```

Comment:
There is no flow where you alloc anon without page. You could just inline page alloc call here and simplify error handling.